### PR TITLE
Refactored ScreenCreditsController to use existing components

### LINKED ISSUE #467: Refactor ScreenCreditController to reuse existing components
Closes #489

Refactor the ScreenCreditController's drawing code to reuse existing drawing components, such as ScreenImage and ScreenText.

### PR DESCRIPTION
**Pull request by Beak-man**
### Description of the Change
- The current `ScreenCreditsController` makes use of its own facilities to render images and text overlays in the viewport. This change enables it to use `ScreenText` and `ScreenImage` instead.
- Makes the `ScreenCreditsController` to operate as an internal layer.

### Why Should This Be In Core?
- It reuses present resources (the two aforementioned classes) and it sets the stage for future enhancements in regards to screen credits as described in the issues documented in epic #489.

### Benefits
 - It greatly simplifies the controller, making it easier to maintain. 
 - Code is reused instead of having redundant rendering facilities. 
 - Sets the stage for enhancing screen credits capabilities, as described in issues related to epic #489.
 - The concept of an internal layer may be useful for other features in the future.

### Potential Drawbacks
- The changes done to enable `ScreenCreditsController` to function as an internal layer, exposed a bug described in #494. `WorldWindow`'s line 1203 was commented out as a temporary workaround.
- Maybe moving `ScreenCreditsController` to the /Layer folder should be considered.
- As for now, a fixed spacing between text and image credits is used. It operates well with current credits as requested by `OpenStreetMapLayer` and `BingTiledImageLayer`, but taller image credits will overlap. This is intended to be addressed in a future PR.

### Applicable Issues
Closes #467

### CODE REVIEW

**Comment by pdavidc** [comment]
@Beak-man A brief scan showed changes to files unrelated to the ScreenCreditController refactoring. Please remove those changes. I'd like to see pull requests that are specific to the subject at hand.

That said, I see you're taking the initiative to correct equality and inequality operators that are at risk of type coercion. This is great! Make those changes in another branch, and submit a separate pull request. You can do this in the future as well.

**Comment by Beak-man** [comment]
Will do @pdavidc !

**Comment by pdavidc** [comment]
@Beak-man this refactoring of ScreenCreditController is wonderfully simple and straightforward. It exemplifies the type of consolidation and simplification we're looking for. 

I have a few comments that I think will polish what you've done, and ought to be simple to implement. I've placed those comments inline.

File: src/render/ScreenCreditController.js [hunk 1]
```diff
@@ -51,41 +52,36 @@ define([
          * @classdesc Collects and displays screen credits.
          */
         var ScreenCreditController = function () {
+            RenderableLayer.call(this, "ScreenCreditController");
```
Comment (pdavidc) [inline src/render/ScreenCreditController.js]: RenderableLayer is a very sensible place to start for a Layer subclass. That said, I don't see how we're benefiting from it. We already manage all the renderable, and override doRender.

File: src/render/ScreenCreditController.js [hunk 2]
```diff
@@ -51,41 +52,36 @@ define([
...
             // Internal. Intentionally not documented.
-            this.imageUrls = [];
+            this.imageCredits = [];
 
             // Internal. Intentionally not documented.
-            this.stringCredits = [];
+            this.textCredits = [];
 
             // Internal. Intentionally not documented.
             this.imageCreditSize = 64;
```
Comment (pdavidc) [inline src/render/ScreenCreditController.js]: This property is no longer used.

File: src/render/ScreenCreditController.js [hunk 3]
```diff
@@ -121,184 +121,34 @@ define([
...
         // Internal use only. Intentionally not documented.
-        ScreenCreditController.prototype.endDrawingCredits = function (dc) {
-            var gl = dc.currentGlContext,
-                program = dc.currentProgram;
-
-            // Clear the vertex attribute state.
-            gl.disableVertexAttribArray(program.vertexPointLocation);
-            gl.disableVertexAttribArray(program.vertexTexCoordLocation);
+        ScreenCreditController.prototype.doRender = function (dc) {
+            var j = 1;
```
Comment (pdavidc) [inline src/render/ScreenCreditController.js]: This property is in need of a more descriptive name. Perhaps creditCount or creditOrdinal.

File: src/render/ScreenCreditController.js [hunk 4]
```diff
@@ -121,184 +121,34 @@ define([
...
-            // Scale the image to fit within a constrained size.
-            imageWidth = activeTexture.imageWidth;
-            imageHeight = activeTexture.imageHeight;
-            if (imageWidth <= this.imageCreditSize && this.imageHeight <= this.imageCreditSize) {
-                scale = 1;
-            } else if (imageWidth >= imageHeight) {
-                scale = this.imageCreditSize / imageWidth;
-            } else {
-                scale = this.imageCreditSize / imageHeight;
+            for (var i = 0; i < this.imageCredits.length; i++) {
```
Comment (pdavidc) [inline src/render/ScreenCreditController.js]: Unintentional O(n) loop. Array.length can be, last I checked, a linear-time operation. The pattern we use throughout WorldWind, though it's not obvious at first glance, is as follows:
`for (var i = 0, len = this.imageCredits.length; i < len; i++) {`

File: test/BasicWorldWindowController.test.js [hunk 5]
```diff
@@ -54,6 +54,7 @@ define([
             return {left: 339.5, top: 225};
         }
     };
+    wwd.layers = [];
```
Comment (pdavidc) [inline test/BasicWorldWindowController.test.js]: Would you please clarify why this change is needed in the tests?

File: src/render/ScreenCreditController.js (hunk 4 above)
Comment (Beak-man) [reply src/render/ScreenCreditController.js]: Interesting. Will do.

File: test/BasicWorldWindowController.test.js (hunk 5 above)
Comment (markpet49) [reply test/BasicWorldWindowController.test.js]: @pdavidc I added this to fix a unit test breakage when we converted ScreenCreditController to a layer.

Comment (markpet49) [reply test/BasicWorldWindowController.test.js]: Also, I moved the Mock WWD creation code to a common routine in the camera branch, so these three changes will all be consolidated.

Comment (pdavidc) [reply test/BasicWorldWindowController.test.js]: Thanks for clarifying, Mark.

File: src/render/ScreenCreditController.js [hunk 6]
```diff
@@ -18,74 +18,58 @@
  */
```
Comment (markpet49) [inline src/render/ScreenCreditController.js]: @pdavidc Should we leave ScreenCreditController in the render folder given that it is now a layer? Given that it is the first of WWW's internal layers, perhaps putting it in the layer folder sends the wrong message.

Comment (pdavidc) [reply src/render/ScreenCreditController.js]: I agree with your thinking @markpet49. At the moment, the layer folder is intended for application focused layers.

File: src/render/ScreenCreditController.js [hunk 7]
```diff
@@ -121,184 +108,33 @@ define([
...
-            program.loadOpacity(gl, this.opacity);
-
-            ScreenCreditController.texCoordMatrix.setToIdentity();
-            ScreenCreditController.texCoordMatrix.multiplyByTextureTransform(activeTexture);
-            program.loadTextureMatrix(gl, ScreenCreditController.texCoordMatrix);
-
-            if (activeTexture.bind(dc)) { // returns false if active texture cannot be bound
-                // Draw the image quad.
-                gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
+            for (i = 0; i < this.textCredits.length; i++) {
```
Comment (pdavidc) [inline src/render/ScreenCreditController.js]: Two problems here:
• Array.length is called every iteration.
• Variable i should be declared at the top of the method. Avoid dependencies on variable hosting.

Comment (Beak-man) [reply src/render/ScreenCreditController.js]: Whoops on that Array.length call.
Both points are corrected now.

**Comment by markpet49** [comment]
@Beak-man It looks like there's an off by one type of error with respect to the Y coordinate placement of the credits. The attached image shows the dev version overlayed with your version semi-transparently. If look at the pixel placement, it looks like the new version is slowly rising upward by increasing amounts on the Y axis. Can you look into this?

**Comment by markpet49** [comment]
credits

**Comment by Beak-man** [comment]
@markpet49 , I'll review the accuracy. Is it worth it, though? The placement in this proposal is hardcoded, as we discussed now, and it doesn't take into account the actual heights of the textures.

**Comment by Beak-man** [comment]
@pdavidc , I'll check that out. It never occurred to me, and I didn't identify how the controller was avoiding this repetition before.

**Comment by markpet49** [comment]
@Beak-man Ah, so the difference is the result of a change in the placement algorithm. If so, then never mind.

**Comment by Beak-man** [comment]
@markpet49 , yes. Previous attempts replicated 1:1 credit placement with how the controller worked before, but that relied on reading the credit ActiveTexture, which then we arranged to avoid. It seems that I'm just adding a pixel with each credit, though. That seems a quick fix.

File: src/render/ScreenCreditController.js [hunk 8]
```diff
@@ -121,184 +115,42 @@ define([
...
-
-            for (var i = 0; i < this.imageUrls.length; i++) {
-                imageHeight = this.drawImageCredit(dc, this.imageUrls[i], imageX, this.margin);
-                if (imageHeight > 0) {
-                    imageX -= (this.margin + this.imageCreditSize);
-                    maxImageHeight = WWMath.max(imageHeight, maxImageHeight);
+            // Verify if text credit is not already in controller, if it is, don't add it.
+            for (var i = 0, len = this.textCredits.length; i < len; i++){
+                if (this.textCredits[i].text === stringCredit){
+                    return 0;
```
Comment (pdavidc) [inline src/render/ScreenCreditController.js]: Just `return;` here.

Comment (Beak-man) [reply src/render/ScreenCreditController.js]: Rght. Bad habits die hard.

**Comment by markpet49** [comment]
Looks good to go....
